### PR TITLE
Fix global transform validity for `Node2D` and `Control`

### LINKED ISSUE #79453: Node2D global_position does not update when set before it's added to the scene tree
### Godot version

4.1 stable, 4.1.1 rc1

### System information

Windows 10, Forward +

### Issue description

When updating the global_position of a Node2D that's outside of the scene tree, the global_position does not update. I came across this issue because when I instantiate a scene, I usually set the instantiated object's properties before adding it to the scene tree. When I tried setting the global_position, I noticed the value seems to stay at (0,0). Interestingly, even though the global_position does not update, the position property does. Is it fine to update a node's (global) position when it's not inside the scene tree? Or should we not do this

Also possibly related to #75676

### Steps to reproduce

```gdscript
extends Node

func _ready() -> void:
	var a = Node2D.new()
	a.global_position = Vector2(1, 1)
	assert(a.global_position == Vector2(1, 1))
```

The above assertion fails on 4.1 stable and 4.1.1 rc1, but succeeds on 4.0.2 and 4.0.3. I didn't test other versions besides these

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by Sauermann**
Set global transform to invalid when changing transform

resolve #79453

Add unit-tests for the bug

### CODE REVIEW

**Review by kleonc — DISMISSED** [review]
This is not a correct fix.

The `global_invalid` flag being initialized to false is correct. A newly created CanvasItem has no parent and its own `transform` is an identity, hence `global_transform` being an identity by default is valid.

Marking global transform as invalid by default would result in some unnecessary one-time updates for actually unchanged global transforms. But it doesn't solve the issue which is: some out of tree modifications (like `Node2D.set_global_position`) result in not marking the global transform as dirty.

E.g. with this PR:
```gdscript
@tool
extends EditorScript

func _run() -> void:
	var a = Node2D.new()
	a.global_position = Vector2(1, 1)
	assert(a.global_position == Vector2(1, 1)) # Passes because of invalid by default.
	a.global_position = Vector2(2, 2)
	assert(a.global_position == Vector2(2, 2)) # Still fails.
```

Also I'm not sure if the `global_invalid` to `global_valid` name change is needed/wanted at all.

**Comment by Sauermann** [comment]
@kleonc Thank you for pointing me in the right direction. I have included your improvements in the PR.

**Comment by Sauermann** [comment]
I have included the invalidation for transform-changing operations in `Control`.

**Comment by Sauermann** [comment]
@kleonc may I ask if your requested changes have been resolved by my updates?

**Comment by kleonc** [comment]
> @kleonc may I ask if your requested changes have been resolved by my updates?

Hmm, seems like it should now be properly invalidating the global transforms. :+1:

However, I'm not sure if these added invalidations don't affect transform notifications in some unwanted manner. Specifically, after brief look at the code, when in the tree `CanvasItem::_notify_transform` calls `_notify_transform(this)`:
https://github.com/godotengine/godot/blob/16a93563bfd3b02ca0a8f6df2026f3a3217f5571/scene/main/canvas_item.h#L154-L162
which early returns in case global transform is invalid:
https://github.com/godotengine/godot/blob/16a93563bfd3b02ca0a8f6df2026f3a3217f5571/scene/main/canvas_item.cpp#L877-L886


Hence I wonder if adding `_invalidate_global_transform();` before `_notify_transform();` (like it's done in many of the changed methods) could have broken some notification propagation or something like that. :thinking: We'll need to ensure nothing is broken I guess (or maybe you've already did that?).

**Comment by Sauermann** [comment]
Thank you for taking the time to guide me through the maze of Transform-caching.
The case, that you have identified was indeed problematic.

With the latest updated I made sure that
- the behavior while `is_inside_tree() == true` is not changed
- while not inside tree, caching status traverses down the scene-tree

File: scene/main/canvas_item.cpp [hunk 1]
```diff
@@ -367,7 +367,12 @@ void CanvasItem::_notification(int p_what) {
 		case NOTIFICATION_WORLD_2D_CHANGED: {
 			_exit_canvas();
 			_enter_canvas();
-		}
+		} break;
+		case NOTIFICATION_PARENTED: {
+			if (!is_inside_tree()) {
+				_notify_transform();
+			}
```
Comment (kleonc) [inline scene/main/canvas_item.cpp:374]: Actually, during `NOTIFICATION_PARENTED` the node is still not in the tree, tree-belonging is being set right after (and it propagates `NOTIFICATION_ENTER_TREE`):
https://github.com/godotengine/godot/blob/16a93563bfd3b02ca0a8f6df2026f3a3217f5571/scene/main/node.cpp#L1378-L1382
Meaning the `if (!is_inside_tree())` check in here is redundant / not needed. But (if I get it right) it also means that on `NOTIFICATION_ENTER_TREE`:

- This is pointless now (it should have been already invalidated at this point (by `_notify_transform()`)):

https://github.com/godotengine/godot/blob/16a93563bfd3b02ca0a8f6df2026f3a3217f5571/scene/main/canvas_item.cpp#L323

- This is still fine / needed ([relevant part should have been skipped in `_notify_transform`](https://github.com/godotengine/godot/blob/16a93563bfd3b02ca0a8f6df2026f3a3217f5571/scene/main/canvas_item.cpp#L890-L901) as the node was not yet in the tree):

https://github.com/godotengine/godot/blob/16a93563bfd3b02ca0a8f6df2026f3a3217f5571/scene/main/canvas_item.cpp#L334-L336

PS Watch out for the minotaur. :thread:

Comment (Sauermann) [reply scene/main/canvas_item.cpp]: Thanks, after digging into these parts of the code I agree, that they are not needed.

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` 6758a7f8c07d1f4c8ec4f052ded6d26402967ebe), it works as expected.

I've also tested a few demo projects (3D Platformer, Control Gallery) and they behave correctly.

**Review by kleonc — APPROVED** [review]
AFAICT it should work fine now. :+1:

**Comment by akien-mga** [comment]
Thanks!

**Comment by YuriSizov** [comment]
Cherry-picked for 4.1.2.

**Pull request merged**